azure-pipelines.yaml

```diff
@@ -57,40 +57,30 @@ steps:
     mkdir test-output
     chmod 777 test-output
   displayName: create test output directory
-  # build test image
+  # build base image
 - task: DockerCompose@0
   inputs:
+    action: Build services
     containerregistrytype: Azure Container Registry
     azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
     azureContainerRegistry: $(azureContainerRegistryFull)
     dockerComposeFile: docker-compose.yaml
     projectName: $(Build.Repository.Name)
     qualifyImageNames: true
-    dockerComposeCommand: -f docker-compose.test.yaml build --no-cache $(imageName)
-  displayName: build tests image
```

Comment:
I would prefer this step to be left in.
There should be a difference between the test image and the base image as the final image will not include dev dependencies.
Rather than remove the step it might be better to change the `build base image` step to build a production image with no dev dependencies.
If we take the step out we'll have to put it back in again when the test and base images diverge.

Reply:
I agree the production build should not include dev dependencies. We have a separate ticket in the backlog to address the Dockerfiles. Whether we need two build stages in the pipeline depends on how we end up structuring the Dockerfile and docker-compose configurations.

For now, I felt that having two stages build the same image was unnecessarily slowing down the pipeline.

Reply:
On the ELM prototype we didn't end up needing two build stages in the pipeline, even after stripping dev tools out of the production build. I think this was because we ran a production-like app container and a separate "watcher" container in development so docker-compose built both images from a single command.

Looking back I think this approach relied on our use of Skaffold to restart the app container on code changes. As we've moved away from Skaffold to docker-compose, we may well end up needing two build stages in the pipeline now.

Reply:
I've restored the separate build steps for test and base images for now. We do need to think more about this when the Dockerfile gets improved (there's already a ticket for that in the backlog). Building the production image from scratch after running tests against a different image carries a risk of divergence so that tests lose value over time. Ideally the production image should be built first with the test build extending from it.

Reply:
The package-lock.json specifies a version, location and integrity hash for every module and each of its dependencies. The install it creates will be the same, every single time.
If the npm install uses package-lock there will be no divergence.